capstone_project/src/main/java/com/google/sps/data/PostService.java

```diff
@@ -103,16 +139,77 @@ public void storePost(HttpServletRequest request) {
    * the new upvote count after increase.
    */
   public long upvotePost(HttpServletRequest request) {
-    Query query = new Query("Post");
+    long postID = Long.parseLong(request.getParameter("id"));
+    Key key = KeyFactory.createKey("Post", postID);
+    Entity postToUpvote = null;
 
-    List<Entity> posts = datastore.prepare(query).asList(FetchOptions.Builder.withDefaults());
+    try {
+      postToUpvote = datastore.get(key);
+    } catch (EntityNotFoundException ok) {
+      // this is okay because by default postToUpvote is null and there is a null check
+    }
 
-    int postID = Integer.parseInt(request.getParameter("id"));
-    Entity postToUpvote = posts.get(postID);
+    if (postToUpvote == null) {
+      return -1;
+    }
 
     long currentUpvotes = (long) postToUpvote.getProperty("upvotes");
     postToUpvote.setProperty("upvotes", ++currentUpvotes);
     datastore.put(postToUpvote);
     return currentUpvotes;
   }
+
+  /**
+   * Sorts a list of entities based on the sort type passed in and returns the sorted list. If the
+   * sort type does not exist, an empty list is returned.
+   */
+  public List<Entity> sortEntities(String sortType, List<Entity> entities) {
+    Comparator<Entity> sortMethod = postSorters.get(sortType);
+    if (sortMethod == null) {
+      return Arrays.asList();
+    }
+
+    Collections.sort(entities, sortMethod);
+    return entities;
+  }
+
+  /** Returns a Map of all comparators. */
+  private void initializeSorters() {
+    postSorters.put("new", this::sortByNew);
+    postSorters.put("top", this::sortByTop);
+    postSorters.put("trending", this::sortByTrending);
+  }
+
+  /** Returns a Comparator that sorts entities by new, highest timestamp to smallest timestamp. */
+  private int sortByNew(Entity first, Entity second) {
+    long firstTime = (long) first.getProperty("timestamp");
+    long secondTime = (long) second.getProperty("timestamp");
+    return Long.compare(secondTime, firstTime);
+  }
+
+  /** Returns a Comparator that sorts by top, most upvotes to least upvotes. */
+  private int sortByTop(Entity first, Entity second) {
+    long firstUpvotes = (long) first.getProperty("upvotes");
+    long secondUpvotes = (long) second.getProperty("upvotes");
+    return Long.compare(secondUpvotes, firstUpvotes);
+  }
+
+  /** Returns a comparator that sorts by trending, highest upvote ratio to lowest. */
+  private int sortByTrending(Entity first, Entity second) {
```

Comment:
I don't quite understand why this and others can't be static? Could you explain a bit more.

Reply:
The function relies on getUpvoteRatio which requires a clock instance. I cannot make non static method calls from a static method since there is no class instance, which means there is no clock instance. line 209 is where the clock instance is used.